src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,765 @@
+        Ok(state)
+    }
+
+    unsafe fn init_compose(&mut self) {
+        let locale = env::var_os("LC_ALL")
+            .and_then(|v| if v.is_empty() { None } else { Some(v) })
+            .or_else(|| env::var_os("LC_CTYPE"))
+            .and_then(|v| if v.is_empty() { None } else { Some(v) })
+            .or_else(|| env::var_os("LANG"))
+            .and_then(|v| if v.is_empty() { None } else { Some(v) })
+            .unwrap_or_else(|| "C".into());
+        let locale = CString::new(locale.into_vec()).unwrap();
+
+        let compose_table = (XKBCH.xkb_compose_table_new_from_locale)(
+            self.xkb_context,
+            locale.as_ptr(),
+            ffi::xkb_compose_compile_flags::XKB_COMPOSE_COMPILE_NO_FLAGS,
+        );
+
+        if compose_table.is_null() {
+            // init of compose table failed, continue without compose
+            return;
+        }
+
+        let compose_state = (XKBCH.xkb_compose_state_new)(
+            compose_table,
+            ffi::xkb_compose_state_flags::XKB_COMPOSE_STATE_NO_FLAGS,
+        );
+
+        if compose_table.is_null() {
+            // init of compose state failed, continue without compose
+            (XKBCH.xkb_compose_table_unref)(compose_table);
+            return;
+        }
+
+        let compose_state_2 = (XKBCH.xkb_compose_state_new)(
+            compose_table,
+            ffi::xkb_compose_state_flags::XKB_COMPOSE_STATE_NO_FLAGS,
+        );
+
+        if compose_state_2.is_null() {
+            // init of compose state failed, continue without compose
+            (XKBCH.xkb_compose_table_unref)(compose_table);
+            (XKBCH.xkb_compose_state_unref)(compose_state);
+            return;
+        }
+
+        self.xkb_compose_table = compose_table;
+        self.xkb_compose_state = compose_state;
+        self.xkb_compose_state_2 = compose_state_2;
+    }
+
+    unsafe fn post_init(&mut self, state: *mut ffi::xkb_state, keymap: *mut ffi::xkb_keymap) {
+        self.xkb_keymap = keymap;
+        self.xkb_state = state;
+        self.mods_state.update_with(state);
+    }
+
+    unsafe fn de_init(&mut self) {
+        (XKBH.xkb_state_unref)(self.xkb_state);
+        self.xkb_state = ptr::null_mut();
+        (XKBH.xkb_keymap_unref)(self.xkb_keymap);
+        self.xkb_keymap = ptr::null_mut();
+    }
+
+    #[cfg(feature = "x11")]
+    pub(crate) unsafe fn init_with_x11_keymap(&mut self) {
+        if !self.xkb_keymap.is_null() {
+            self.de_init();
+        }
+
+        // TODO: Support keyboards other than the "virtual core keyboard device".
+        self.core_keyboard_id = (XKBXH.xkb_x11_get_core_keyboard_device_id)(self.xcb_connection);
+        let keymap = (XKBXH.xkb_x11_keymap_new_from_device)(
+            self.xkb_context,
+            self.xcb_connection,
+            self.core_keyboard_id,
+            xkbcommon_dl::xkb_keymap_compile_flags::XKB_KEYMAP_COMPILE_NO_FLAGS,
+        );
+        assert_ne!(keymap, ptr::null_mut());
```

Comment:
Do we really need assert like that? I'd assume it's when keymap can't be applied, but should at least log or something like that what is failed.